### ISSUE TITLE
Dead kernel when computing neighbours 

### ISSUE THREAD

**Issue by ccruizm** [theislab/scvelo#198]
Good day,

I am having a problem running `scVelo`. I am doing all the preprocessing steps in `Seurat` (integrated two different samples), and then, I exported the loom file and read it into `scVelo` successfully. Then I combined the loom files obtained after running `velocyto` into one using `loompy.combine`. I can merge both objects using `scv.utils.merge` without losing any cells, and the `AnnData` object seems ok to me (I am new using python for single-cell RNA processing so that I may be wrong). However, the problem comes when I want to run `scv.tl.velocity`. The command will start the normalization, but once it starts to compute the neighbors, it kills the kernel. 

I am working on an HPC and have tried increasing the number of cores and memory (12 cores and 256GB), thinking that that could be the problem, but still, the problem occurs again. the dataset is not huge (`AnnData object with n_obs × n_vars = 14542 × 17586`), so I do not have any clue where the problem is. I have run the pipeline with non-merged samples, and it finishes with no issues.

I am quite puzzled about where the problem could be. I am sorry if this is a trivial question. I hope you may have some ideas about how to solve it. 

Thanks in advance.

**Comment by ccruizm** [comment]
Yes, I am running ` scvelo 0.2.0 (python 3.8.2)`. I have tried also running the commands you mentioned before and kills the kernel when I start running `scv.pp.moments`. I tried using different methods, instead of the default, but every time, it takes a few seconds and then kills and restarts the kernel.

**Comment by VolkerBergen** [comment]
Could you please try to run neighbors via `hnswlib`, i.e., 
`pip install pybind11 hnswlib` and then `scv.pp.moments(adata, method='hnsw')`.

Gives essentially the same neighbor graph, but is bit more efficient.

**Comment by ccruizm** [comment]
It still has the same problem. I decided to test it on another HPC and there it ran with no issues. I do not have any idea what the problem is on the other server. Maybe something in the conda env or the scheduling system. 

Thanks for the quick replies and help!

**Comment by VolkerBergen** [comment]
alright, maybe upgrading all potentially related requirements solves it
`pip install --upgrade anndata scanpy numpy scipy scikit-learn pandas matplotlib umap-learn h5py statsmodels`

**Comment by picciama** [comment]
@VolkerBergen:
I've go the same problem and unfortunately it cannot be solved using your suggestions. 
Versions:
>scvelo==0.2.1  scanpy==1.5.1  anndata==0.7.3  loompy==3.0.6  numpy==1.18.4  scipy==1.4.1  matplotlib==3.2.1  sklearn==0.23.1  pandas==1.0.4 

When running it in a terminal instead of jupyter, I get a "Speicherzugriffsfehler" (translates to SEGFAULT if I remember correctly) before the script crashes. 
I use an anndata object that already contains neighbors because PCA and UMAP were computed previously. 
Is there a way to compute moments without recomputing the neighbors?
Or existing neighbors alone a source of problems?

So I tried calling moments on just the unspliced data before merging it with the existing spliced data. I has the same effect.

@ccruizm I wonder if it worked for you?

**Comment by VolkerBergen** [comment]
Thanks for raising/reopening this. It seems to be related to https://github.com/theislab/anndata/issues/380. 

Could you please try this once more with the lates source version of anndata?
`pip install git+https://github.com/theislab/anndata`

**Comment by picciama** [comment]
>anndata==0.7.4.dev3+gaf24e3e
doesn't help either. For now I'm using steady state without calculating moments and instead take the once from my existing PCA/neighbors calculation. We're going to debug it and update here with more details then.

**Comment by VolkerBergen** [comment]
As workaround for now you can store the distances as
```
adata.uns['neighbors'] = dict(
    distances = sparse_matrix_with_distances, 
    params = {"n_neighbors": 30, "n_pcs": 30}
)
```
This way, all relevant information is stored to avoid recomputing the graph.

@flying-sheep The `obsp` memory issue doesn't seem to be fully solved yet.

**Comment by picciama** [comment]
I found my problem and it seems unrelated to `obsp`:
I can calculate neighbors when calling the `moments` only once. The second time it fails with the dead kernel issue. I noticed that within the executed cell in my notebook,  `scv.logging.print_version()` is called. As soon as I removed that line, I could calculate moments/neighbors repeatedly without any problem. It seems as if `scv.logging.print_version()` must not be called again after `neighbors` was called at least once.
In my case, this was an unfortunate combination of a for loop and multiple anndata objects.

Steps to reproduce:
1. first call `scv.pp.moments()`
2. then call `scv.logging.print_version()`
3. call `scv.pp.moments()` again, either on the same or a different object <- kernel death

The line the kernel dies is this one here: 
https://github.com/theislab/scvelo/blob/4395ae0e4ecf0452a100488692205bd054e3c009/scvelo/preprocessing/neighbors.py#L144

I checked and observed this behavior with scvelo from 0.1.9 to the lastest dev version.
Though this special case I report is likely never to occur, there might be a general problem with the logging mechanism. However, the issue can be settled in my opinion.

**Comment by KoichiHashikawa** [comment]
@VolkerBergen 
I have the same issue when running scv.pp.moments
Is the following code the replacement for scv.pp.moments? what is sparse_matrix_with_distances?
adata.uns['neighbors'] = dict( distances = sparse_matrix_with_distances,   params = {"n_neighbors": 30, "n_pcs": 30}
)

**Comment by VolkerBergen** [comment]
@picciama strange indeed. Do you get the same error when trying different methods for computing neighbors, i.e. `scv.pp.moments(adata, method="hnsw")` and `scv.pp.moments(adata, method="sklearn")`?

**Comment by VolkerBergen** [comment]
@KoichiHashikawa replacement for computing neighbors. sparse_matrix_with_distances is a matrix of all cell-to-cell distances (usually obtained as Euclidean distance in pca space), stored as a [sparse matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html). Do you also get a dead kernel when running `scv.pp.neighbors(adata)`?

**Comment by picciama** [comment]
> @picciama strange indeed. Do you get the same error when trying different methods for computing neighbors, i.e. `scv.pp.moments(adata, method="hnsw")` and `scv.pp.moments(adata, method="sklearn")`?

Yes, the kernel dies (jupyter) or SEGFAULT (terminal) in the following lines:
**`method="sklearn"`:**
https://github.com/theislab/scvelo/blob/a283037ea5043d8ecdbd0ee26e13c4f150126cb8/scvelo/preprocessing/neighbors.py#L128
**`method="hnsw"`:**
https://github.com/theislab/scvelo/blob/a283037ea5043d8ecdbd0ee26e13c4f150126cb8/scvelo/preprocessing/neighbors.py#L226
**`method="umap"`:**
https://github.com/theislab/scvelo/blob/a283037ea5043d8ecdbd0ee26e13c4f150126cb8/scvelo/preprocessing/neighbors.py#L147
But `umap` can be further traced to this line in scanpy (tested on latest dev branch):
https://github.com/theislab/scanpy/blob/e33a2f338a4c63c794fe4bc5e03471d7c487cada/scanpy/neighbors/__init__.py#L740

So in all three cases it is the `_compute_connectivities_umap()` method that is called. This method isn't entered, I checked with printouts. SEGFAULT is already thrown during the respective method call. Hope it helps :)

**Comment by VolkerBergen** [comment]
That's very helpful indeed. So it traces back to umap's connectivities which is using numba. Could you try running with upgraded umap-learn and upgraded numba?

**Comment by KoichiHashikawa** [comment]
scv.pp.neighbors(adata) killed the kernel as well. It seems so far that anything related to computing neighbors kill kernel when using my own data or using the build in pancreas data.

**Comment by VolkerBergen** [comment]
@KoichiHashikawa even with updated numba and umap-learn?
`pip install -U numba umap-learn`

**Comment by picciama** [comment]
> That's very helpful indeed. So it traces back to umap's connectivities which is using numba. Could you try running with upgraded umap-learn and upgraded numba?

That doesn't do the trick unfortunately. But it works in my case if I remove `scv.logging.print_version()`

My versions as of now:
>scvelo==0.2.2.dev6+g0eafcc1.d20200613  scanpy==1.5.2.dev5+ge5d246aa.d20200613  anndata==0.7.3  loompy==2.0.17  numpy==1.18.4  scipy==1.4.1  matplotlib==3.2.1  sklearn==0.23.1  pandas==1.0.4 numba==0.5.0 umap-learn==0.4.4

**Comment by VolkerBergen** [comment]
Hmm.. Possibly related to https://github.com/lmcinnes/umap/issues/416. 
Try again running after `pip install pynndescent`.

**Comment by picciama** [comment]
> Hmm.. Possibly related to [lmcinnes/umap#416](https://github.com/lmcinnes/umap/issues/416).
> Try again running after `pip install pynndescent`.

You got it! Now it's working for umap, hnsw and sklearn even if calling `print_version()` again.

**Comment by VolkerBergen** [comment]
Wonderful! Will follow up with `umap-learn` how to avoid this.

@KoichiHashikawa resolved for you as well?

**Comment by KoichiHashikawa** [comment]
Thanks for the advice!
I tried pip install pynndescent and pip install -U numba umap-learn.
But I still did not have luck and kernel was killed. My case seems that Kernel is killed when computing neighbors.

**Comment by VolkerBergen** [comment]
Can you find out at which code line it stops when you interrupt the kernel before it crashes?

**Comment by KoichiHashikawa** [comment]
How do I test which line causing error? I have been primarily using R language and not familiar yet with Python, sorry.
In the meantime, I have set up an new conda environment with python 3.8.3 where I installed the latest developmental scvelo along with pynndescent and upgraded other dependencies. I replicated an issue in that environment as well --- scv.pp.neighbors or scv.pp.moments terminated the kernel.

**Comment by picciama** [comment]
@KoichiHashikawa You may wish to `git clone` scvelo and scanpy. Then `pip install -e <dir>` whereby `dir` is the directory you cloned the two into. Now all changes in the code of these two repos are reflected when you import the packages. You can e.g. write a simple `print()` and see if it is still printed to stdout or if the kernel dies before you see the output.
I think you should start your search right before the functions I've linked above are called.
A good start is the `neighbors()` function in scvelo/preprocessing/neighbors.py

**Comment by KoichiHashikawa** [comment]
@picciama 
Thanks a lot for the advice. I installed scvelo and scanpy via git clone, but did not have a luck.

I checked lines causing kernel dead by inserting print(1),print(2),,, in neighbors() function and found that as the same as you did that this line " neighbors.compute_neighbors( n_neighbors=n_neighbors, knn=knn, n_pcs=n_pcs,        method=method, use_rep=None if use_rep == "X_pca" else use_rep, random_state=random_state,           metric=metric,  metric_kwds=metric_kwds, write_knn_indices=True, )" killed the kernel.

**Comment by KoichiHashikawa** [comment]
I just read back the thread and removed scv.logging.print_version() and now it ran.  
Thanks so much @picciama @VolkerBergen .

**Comment by VolkerBergen** [comment]
`scv.logging.print_version()` causing problems is a really mysterious.. Would like to understand what the issue is with that function - and how that'd be related to `neighbors`. Could you comment out [this line](https://github.com/theislab/scvelo/blob/f43839635c2f90acc53ed0a59abb22d37dcf800c/scvelo/logging.py#L209) and see whether that runs then smoothly? 

And if you use `scv.pp.moments(adata, method="sklearn")` where does it break?

**Comment by whosnothungry** [comment]
> I found my problem and it seems unrelated to `obsp`:
> I can calculate neighbors when calling `moments()` only once. The second time it fails with the dead kernel issue. I noticed that within the executed cell in my notebook, `scv.logging.print_version()` is called. As soon as I removed that line, I could calculate moments/neighbors repeatedly without any problem. It seems as if `scv.logging.print_version()` must not be called again after `neighbors()` was called at least once.
> In my case, this was an unfortunate combination of a for loop and multiple anndata objects.
> 
> Steps to reproduce:
> 
> 1. first call `scv.pp.moments()`
> 2. then call `scv.logging.print_version()`
> 3. call `scv.pp.moments()` again, either on the same or a different object <- kernel death
> 
> The line the kernel dies is this one here:
> https://github.com/theislab/scvelo/blob/4395ae0e4ecf0452a100488692205bd054e3c009/scvelo/preprocessing/neighbors.py#L144
> 
> I checked and observed this behavior with scvelo from 0.1.9 to the lastest dev version.
> Though this special case I report is likely never to occur, there might be a general problem with the logging mechanism. However, the issue can be settled in my opinion.

Same solution
